clang/lib/CodeGen/CGCall.cpp

```diff
@@ -5017,8 +5017,9 @@ RValue CodeGenFunction::EmitCall(const CGFunctionInfo &CallInfo,
         EmitNounwindRuntimeCall(Fn);
       }
     }
-
-    EmitUnreachable(Loc);
+    // Do not emit unreachable instruction for Intel NEO driver and Gen "X" GPUs
+    if (getTarget().getTriple().getSubArch() != llvm::Triple::SPIRSubArch_gen)
```

Comment:
I don't get why NEO driver can't handle this.
It seems that SPIR-V has unreachable instruction as well.

Reply:
> SPIR-V does not support semantic of `noreturn` functions; there is no (portable) way to terminate an invocation or jump out of a function.

How is this related to removing "unreachable" instruction from LLVM IR?

As noted by @Fznamznon, SPIR-V has ["unreachable"](https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#OpUnreachable) operation, which behavior is undefined by the spec. Could you elaborate more on why you think that removing "unreachable" is the right way to fix the original problem? I don't see the connection.

Reply:
If a function is marked as `noreturn`, it has to either terminate, throw or longjmp. Otherwise the behavior is undefined, so clang emits an unreachable instruction after a call to a noreturn function.

There is no way to "not return" in SPIR-V, so any call to a `noreturn` function is a UB. Currently this UB effectively marks the whole basic block unreachable. The proposed patch changes this UB to ignore `noreturn`. This gives more room to handle existing code that has `noreturn` functions.

Reply:
Why can't we leave SPIR-V unreachable and let GPU back-end to handle UB? If I understand correctly SPIR-V already ignores `noreturn` attribute.

Reply:
Frontend has more information and can distinguish `noreturn` attribute from other uses of unreachable (the corresponding builtin, for instance). 
Also, when optimizations are enabled in the frontend, optimizer may remove code related to an unreachable.

Reply:
@asavonic I think this change is causing the original test you added (devicelib/assert.cpp) to fail.
[assert.cpp](http://ci.llvm.intel.com:8010/#/builders/2/builds/2826)
`Basic Block in function '_ZZZ11simple_vaddIiLm3EEvRKSt5arrayIT_XT0_EES4_RS2_ENKUlRN2cl4sycl7handlerEE_clES9_ENKUlNS7_2idILi1EEEE_clESC_' does not have terminator!`

Can you and @bader decide and suggest what's the best way to address the original issue of this PR?

Reply:
> @asavonic I think this change is causing the original test you added (devicelib/assert.cpp) to fail.
> [assert.cpp](http://ci.llvm.intel.com:8010/#/builders/2/builds/2826)
> `Basic Block in function '_ZZZ11simple_vaddIiLm3EEvRKSt5arrayIT_XT0_EES4_RS2_ENKUlRN2cl4sycl7handlerEE_clES9_ENKUlNS7_2idILi1EEEE_clESC_' does not have terminator!`

Sounds like we need to skip the entire block at `if (CI->doesNotReturn())`.